src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10949,99 +10946,143 @@ private ExpressionSyntax ParseSubExpressionCore(Precedence precedence)
             // parse out a missing name node for the expression, and keep on going
             var tk = this.CurrentToken.Kind;
             if (IsInvalidSubExpression(tk))
-            {
                 return this.AddError(this.CreateMissingIdentifierName(), ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
-            }
 
-            // Parse a left operand -- possibly preceded by a unary operator.
-            if (IsExpectedPrefixUnaryOperator(tk))
-            {
-                var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
-                newPrecedence = GetPrecedence(opKind);
-                var opToken = this.EatToken();
-                var operand = this.ParseSubExpression(newPrecedence);
-                leftOperand = _syntaxFactory.PrefixUnaryExpression(opKind, opToken, operand);
-            }
-            else if (tk == SyntaxKind.DotDotToken)
+            return ParseExpressionContinued(parseUnaryOrPrimaryExpression(precedence), precedence);
+
+            // Parses out a unary expression, or something with higher precedence (cast, addressof, primary).
+            ExpressionSyntax parseUnaryOrPrimaryExpression(Precedence precedence)
             {
-                // Operator ".." here can either be a prefix unary operator or a stand alone empty range:
-                var opToken = this.EatToken();
-                newPrecedence = GetPrecedence(SyntaxKind.RangeExpression);
+                // Parse a left operand -- possibly preceded by a unary operator.
+                if (IsExpectedPrefixUnaryOperator(tk))
+                {
+                    var opKind = SyntaxFacts.GetPrefixUnaryExpression(tk);
+                    return _syntaxFactory.PrefixUnaryExpression(
+                        opKind,
+                        this.EatToken(),
+                        this.ParseSubExpression(GetPrecedence(opKind)));
```

Comment:
moved to a model where instead of assinging to a variable (and then having to see what happens to it), we have a local function which just returns the value in question.  Making it much easier to see that this is just responsible for parsing out that construct and *nothing else*.

also, moved to the much clearer parsing pattern we try to use int eh rest of hte parser where the production `A -> B C D` is parsed as `return A(ParseB(), ParseC(), ParseD())`.  this lowers the number of locals to keep track of and ensures not accidentally using them for an innapropriate purpose.